### PR TITLE
Use new text-buffer version with a more efficient DisplayLayer

### PR DESCRIPTION
**Pull request by maxbrunsfeld**
Depends on https://github.com/atom/text-buffer/pull/185. See that PR for a more thorough description.

Fixes https://github.com/atom/atom/issues/8864
Fixes https://github.com/atom/atom/issues/12391

This removes some restrictions that we previously enforced when editing large files; you can now use folds and soft wraps. The only feature that remains disabled for > 10MB files is syntax highlighting.

Large files also now open somewhat faster. Here are some benchmark results from Circle CI [before](https://circleci.com/gh/atom/atom/2204?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) and [after](https://circleci.com/gh/atom/atom/2205?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) this PR:

benchmarks

There are lot of noise in these results of course, but opening a 10MB file is consistently ~200ms faster.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -65,7 +65,7 @@
     "sinon": "1.17.4",
     "source-map-support": "^0.3.2",
     "temp": "0.8.1",
-    "text-buffer": "9.4.3",
+    "text-buffer": "v10.0.0-4",
```
Comment (50Wliu) [inline package.json]: This should be 10.0.0-4 to be consistent, right?

Comment (maxbrunsfeld) [inline package.json]: Oh yeah, thanks. I'll fix this when I bump it to `10.0.0`.

**Comment by Ben3eeE** [comment]
Really loving this :racehorse: :heart_eyes_cat: will test it out more with the new `spell-check` and `whitespace` changes tonight(Late American morning). But don't wait for me if you want to go ahead and :shipit: 

@ungb @maxbrunsfeld Want to drop https://github.com/atom/atom/issues/11690 here since it is an uncaught that reproduces with large files. We don't have a file to repro or any steps other than open a large file but I believe it can be worth some :eyes: given these changes.

**Comment by maxbrunsfeld** [comment]
@Ben3eeE I believe #11690 is due to a separate limitation: we store the entire contents of a file in JavaScript string, and strings have a hard limit on their lengths. This PR does not fix that issue.

**Comment by Stanzilla** [comment]
@maxbrunsfeld This changes nothing for https://github.com/atom/atom/issues/1667, right?

**Comment by maxbrunsfeld** [comment]
> This changes nothing for #1667, right?

Right. It does not address that issue.

**Comment by alexandernst** [comment]
@maxbrunsfeld Is there an issue which we can track for ast parsing/syntax highlight? Where can we expect such a feature to be ready?

**Pull request merged**